crates/red_knot_python_semantic/src/types.rs

```diff
@@ -667,6 +667,55 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Return true if this type is non-empty and all inhabitants of this type compare equal.
+    pub(crate) fn is_single_valued(self, db: &'db dyn Db) -> bool {
+        match self {
+            Type::None
+            | Type::Function(..)
+            | Type::Module(..)
+            | Type::Class(..)
+            | Type::IntLiteral(..)
+            | Type::BooleanLiteral(..)
+            | Type::StringLiteral(..)
+            | Type::BytesLiteral(..) => true,
+
+            Type::Tuple(tuple) => tuple
+                .elements(db)
+                .iter()
+                .all(|elem| elem.is_single_valued(db)),
+
+            Type::Instance(class_type) => match class_type.known(db) {
+                Some(KnownClass::NoneType) => true,
+                Some(
+                    KnownClass::Bool
+                    | KnownClass::Object
+                    | KnownClass::Bytes
+                    | KnownClass::Type
+                    | KnownClass::Int
+                    | KnownClass::Float
+                    | KnownClass::Str
+                    | KnownClass::List
+                    | KnownClass::Tuple
+                    | KnownClass::Set
+                    | KnownClass::Dict
+                    | KnownClass::GenericAlias
+                    | KnownClass::ModuleType
+                    | KnownClass::FunctionType,
+                ) => false,
+                None => false,
+            },
+
+            Type::Any
+            | Type::Never
+            | Type::Unknown
+            | Type::Unbound
+            | Type::Todo
+            | Type::Union(..)
+            | Type::Intersection(..)
```

Comment:
I think an intersection would be single-valued if at least one type in its positive side is single-valued.

But I think this can only occur, given intersection simplification, if there are two different single-valued types that are overlapping (not disjoint) but neither is a subtype of the other. If they are disjoint, or if one is a subtype of the other, then they can't occur in the same intersection without one or both of them disappearing (unless they are disjoint and both on the negative side, which wouldn't make the intersection single-valued).

If there _were_ two different single-valued types that are not disjoint, that would imply that all inhabitants of both types are in the same equivalence class. I have a hard time imagining that we'd ever introduce such a pair of types; it certainly doesn't occur for any of our current single-valued types. And even harder to imagine that it could occur without one being a subtype of the other.

So I don't think this case is important :)